### PR TITLE
Migrating PaintWorklet layout tests to wpt

### PR DESCRIPTION
**Pull request by xidachen**
We have a set of layout tests under chromium repo for PaintWorklet. Before shipping PaintWorklet, we would want to migrate the tests wpt.

The set of tests in this pull request are all ref tests. More tests will come soon.

Please review. @bfgeek @foolip

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239945709)

# Firefox (nightly) #
Testing web-platform-tests at revision 114231ef27d73ccd458480ba25fc9e9654d7fae1
Using browser at version BuildID 20170605100313; SourceStamp 275588f4d852d7dc183a9dcc70a311413dc7a063
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>19 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-alpha.html">/css-paint-api/background-image-alpha.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-multiple.html">/css-paint-api/background-image-multiple.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-tiled.html">/css-paint-api/background-image-tiled.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-constructor-error.html">/css-paint-api/invalid-image-constructor-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-paint-error.html">/css-paint-api/invalid-image-paint-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-pending-script.html">/css-paint-api/invalid-image-pending-script.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/overdraw.html">/css-paint-api/overdraw.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-arguments.html">/css-paint-api/paint-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-function-arguments.html">/css-paint-api/paint-function-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-composite.html">/css-paint-api/paint2d-composite.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-filter.html">/css-paint-api/paint2d-filter.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-gradient.html">/css-paint-api/paint2d-gradient.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-image.html">/css-paint-api/paint2d-image.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-paths.html">/css-paint-api/paint2d-paths.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-rects.html">/css-paint-api/paint2d-rects.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-shadows.html">/css-paint-api/paint2d-shadows.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-transform.html">/css-paint-api/paint2d-transform.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-after-load.html">/css-paint-api/valid-image-after-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-before-load.html">/css-paint-api/valid-image-before-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239945712)

# Sauce (safari) #
Testing web-platform-tests at revision 114231ef27d73ccd458480ba25fc9e9654d7fae1
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>19 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-alpha.html">/css-paint-api/background-image-alpha.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-multiple.html">/css-paint-api/background-image-multiple.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-tiled.html">/css-paint-api/background-image-tiled.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-constructor-error.html">/css-paint-api/invalid-image-constructor-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-paint-error.html">/css-paint-api/invalid-image-paint-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-pending-script.html">/css-paint-api/invalid-image-pending-script.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/overdraw.html">/css-paint-api/overdraw.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-arguments.html">/css-paint-api/paint-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-function-arguments.html">/css-paint-api/paint-function-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-composite.html">/css-paint-api/paint2d-composite.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-filter.html">/css-paint-api/paint2d-filter.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-gradient.html">/css-paint-api/paint2d-gradient.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-image.html">/css-paint-api/paint2d-image.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-paths.html">/css-paint-api/paint2d-paths.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-rects.html">/css-paint-api/paint2d-rects.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-shadows.html">/css-paint-api/paint2d-shadows.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-transform.html">/css-paint-api/paint2d-transform.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-after-load.html">/css-paint-api/valid-image-after-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-before-load.html">/css-paint-api/valid-image-before-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239945725)

# Chrome (unstable) #
Testing web-platform-tests at revision 114231ef27d73ccd458480ba25fc9e9654d7fae1
Using browser at version 60.0.3112.10 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>19 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-alpha.html">/css-paint-api/background-image-alpha.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-multiple.html">/css-paint-api/background-image-multiple.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/background-image-tiled.html">/css-paint-api/background-image-tiled.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-constructor-error.html">/css-paint-api/invalid-image-constructor-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-paint-error.html">/css-paint-api/invalid-image-paint-error.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/invalid-image-pending-script.html">/css-paint-api/invalid-image-pending-script.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/overdraw.html">/css-paint-api/overdraw.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-arguments.html">/css-paint-api/paint-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint-function-arguments.html">/css-paint-api/paint-function-arguments.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-composite.html">/css-paint-api/paint2d-composite.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-filter.html">/css-paint-api/paint2d-filter.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-gradient.html">/css-paint-api/paint2d-gradient.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-image.html">/css-paint-api/paint2d-image.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-paths.html">/css-paint-api/paint2d-paths.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-rects.html">/css-paint-api/paint2d-rects.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-shadows.html">/css-paint-api/paint2d-shadows.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/paint2d-transform.html">/css-paint-api/paint2d-transform.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-after-load.html">/css-paint-api/valid-image-after-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6163/css-paint-api/valid-image-before-load.html">/css-paint-api/valid-image-before-load.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239945726)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 114231ef27d73ccd458480ba25fc9e9654d7fae1
Using browser at version 14.14393
Starting 10 test iterations

**Comment by foolip** [comment]
@xidachen, working directly in this repository is just fine, but if you'd like, you can move them, deleting the old tests, in a Chromium CL directly. It saves you a bit of work in total.

I'll review this though.

File: common/css-paint-tests.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// We call rAF and in the second frame we compare the result with the ref.html.
+function importPaintWorkletAndTerminateTestAfterAsyncPaint(code) {
+    if (window.testRunner) {
+        testRunner.waitUntilDone();
```
Comment (foolip) [inline common/css-paint-tests.js]: These are WebKit/Blink internals and shouldn't be included in wpt. The `class="reftest-wait"` mechanism is probably what's needed, so common/reftest-wait.js for a script wrapping that.

File: common/css-paint-tests.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    var blob = new Blob([code], {type: 'text/javascript'});
+    var frame_cnt = 0;
+    paintWorklet.addModule(URL.createObjectURL(blob)).then(function() {
+        requestAnimationFrame(function() {
+            takeScreenshot(frame_cnt);
+        });
+    });
+}
+
+function takeScreenshot(frame_cnt) {
```
Comment (foolip) [inline common/css-paint-tests.js]: How about introducing a `waitForAnimationFrames(n)` or similar, and using, `waitForAnimationFrames(n).then(takeScreenshot)`, where `takeScreenshot` comes from common/reftest-wait.js?

File: css-paint-api/background-image-alpha-ref.html [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  <div class="container">
+    <canvas id ="opaque" width="100" height="100"></canvas>
+  </div>
+  <div class="container">
+    <canvas id ="nonopaque" width="100" height="100"></canvas>
+  </div>
+</div>
+
+<script>
+function drawCanvas(canvasID, hasAlpha) {
```
Comment (foolip) [inline css-paint-api/background-image-alpha-ref.html]: This seems a bit elaborate for the reference file, would divs with `box-sizing: border-box; border: 4px solid blue` achieve the same result?

File: css-paint-api/background-image-alpha.html [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  background-image: paint(nonOpaque);
+}
+
+#background {
+  background-color: yellow;
+  display: inline-block;
+}
+</style>
+
+<script src="../common/css-paint-tests.js"></script>
```
Comment (foolip) [inline css-paint-api/background-image-alpha.html]: Use absolute paths for /common/*

**Review by foolip — CHANGES_REQUESTED** [review]
I've only skimmed the changes since these are converted from existing tests that have already been reviewed.

**Comment by xidachen** [comment]
@foolip thanks for pointing it out. I didn't even know that it is a two way sync between chromium and wpt. For the rest of the tests, I will do that in a chromium CL instead of doing it in two steps :)

**Comment by foolip** [comment]
I will be away for the rest of the week so someone else needs to follow up, or you could move the review to Chromium. So that my review doesn't block merging, I'll approve it now, treat as LGTM % nits.

**Review by foolip — APPROVED** [review]

File: css-paint-api/background-image-alpha.html [hunk 5]
```diff
@@ -21,7 +21,7 @@
 }
 </style>
 
-<script src="../common/css-paint-tests.js"></script>
+<script src="/common/css-paint-tests.js"></script>
```
Comment (foolip) [inline css-paint-api/background-image-alpha.html]: Do the tests still work, or do they all need to include /common/reftest-wait.js?

Comment (xidachen) [inline css-paint-api/background-image-alpha.html]: Sorry, I am writing another patch, will be up in a few mins.

**Comment by xidachen** [comment]
@foolip I have made changes, tested locally and all of them passed.

**Pull request closed without merging**